*** ../binutils-2.24.orig/bfd/elfnn-aarch64.c	2013-12-17 11:16:28.723807381 +0000
--- bfd/elfnn-aarch64.c	2013-12-17 11:18:13.517804067 +0000
*************** _aarch64_elf_section_data;
*** 1679,1686 ****
  #define elf_aarch64_section_data(sec) \
    ((_aarch64_elf_section_data *) elf_section_data (sec))
  
! /* The size of the thread control block.  */
! #define TCB_SIZE	16
  
  struct elf_aarch64_local_symbol
  {
--- 1679,1686 ----
  #define elf_aarch64_section_data(sec) \
    ((_aarch64_elf_section_data *) elf_section_data (sec))
  
! /* The size of the thread control block which is defined to be two pointers.  */
! #define TCB_SIZE	(ARCH_SIZE/8)*2
  
  struct elf_aarch64_local_symbol
  {
