diff -rup binutils.orig/gas/testsuite/gas/ppc/vsx4.d binutils-2.35/gas/testsuite/gas/ppc/vsx4.d
--- binutils.orig/gas/testsuite/gas/ppc/vsx4.d	2020-08-24 13:52:14.230454823 +0100
+++ binutils-2.35/gas/testsuite/gas/ppc/vsx4.d	2020-08-24 13:54:15.331718297 +0100
@@ -7,6 +7,6 @@
 Disassembly of section \.text:
 
 0+0 <vsx4>:
-.*:	(f0 50 6f 6f|6f 6f 50 f0) 	xvcvbf16sp vs34,vs45
+.*:	(f0 50 6f 6f|6f 6f 50 f0) 	xvcvbf16spn vs34,vs45
 .*:	(f1 f1 27 6f|6f 27 f1 f1) 	xvcvspbf16 vs47,vs36
 #pass
diff -rup binutils.orig/gas/testsuite/gas/ppc/vsx4.s binutils-2.35/gas/testsuite/gas/ppc/vsx4.s
--- binutils.orig/gas/testsuite/gas/ppc/vsx4.s	2020-08-24 13:52:14.230454823 +0100
+++ binutils-2.35/gas/testsuite/gas/ppc/vsx4.s	2020-08-24 13:54:27.779642594 +0100
@@ -1,4 +1,4 @@
 	.text
 vsx4:
-	xvcvbf16sp	34,45
+	xvcvbf16spn	34,45
 	xvcvspbf16	47,36
diff -rup binutils.orig/opcodes/ppc-opc.c binutils-2.35/opcodes/ppc-opc.c
--- binutils.orig/opcodes/ppc-opc.c	2020-08-24 13:52:13.709457991 +0100
+++ binutils-2.35/opcodes/ppc-opc.c	2020-08-24 13:53:46.810891761 +0100
@@ -8306,7 +8306,7 @@ const struct powerpc_opcode powerpc_opco
 {"xvxexpsp",	XX2VA(60,475,8),XX2_MASK,    PPCVSX3,	PPCVLE,		{XT6, XB6}},
 {"xvxsigsp",	XX2VA(60,475,9),XX2_MASK,    PPCVSX3,	PPCVLE,		{XT6, XB6}},
 {"xxbrw",	XX2VA(60,475,15),XX2_MASK,   PPCVSX3,	PPCVLE,		{XT6, XB6}},
-{"xvcvbf16sp",	XX2VA(60,475,16),XX2_MASK,   PPCVSX4,	PPCVLE,		{XT6, XB6}},
+{"xvcvbf16spn",	XX2VA(60,475,16),XX2_MASK,   PPCVSX4,	PPCVLE,		{XT6, XB6}},
 {"xvcvspbf16",	XX2VA(60,475,17),XX2_MASK,   PPCVSX4,	PPCVLE,		{XT6, XB6}},
 {"xxbrd",	XX2VA(60,475,23),XX2_MASK,   PPCVSX3,	PPCVLE,		{XT6, XB6}},
 {"xvcvhpsp",	XX2VA(60,475,24),XX2_MASK,   PPCVSX3,	PPCVLE,		{XT6, XB6}},
