diff -rup binutils-2.27.orig/bfd/elf32-arm.c binutils-2.27/bfd/elf32-arm.c
--- binutils-2.27.orig/bfd/elf32-arm.c	2017-02-15 17:25:01.331124382 +0000
+++ binutils-2.27/bfd/elf32-arm.c	2017-02-15 17:25:32.044613743 +0000
@@ -18172,7 +18172,7 @@ elf32_arm_count_additional_relocs (asect
 {
   struct _arm_elf_section_data *arm_data;
   arm_data = get_arm_elf_section_data (sec);
-  return arm_data->additional_reloc_count;
+  return arm_data == NULL ? 0 : arm_data->additional_reloc_count;
 }
 
 /* Called to set the sh_flags, sh_link and sh_info fields of OSECTION which
